bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java

```diff
@@ -0,0 +1,278 @@
+        }
+
+        return timeZone;
+    }
+
+    @Override
+    public void setTimeZone(TimeZone timeZone) throws IllegalArgumentException, ParseException {
+        if (timeZone == null) {
+            throw new IllegalArgumentException("The time zone must not be null");
+        }
+        this.timeZone = timeZone;
+        parseExpression(expression);
+    }
+
+    @Override
+    public String getExpression() {
+        return expression;
+    }
+
+    @Override
+    public void setExpression(String expression) throws ParseException {
+        this.expression = expression;
+        parseExpression(expression);
+    }
+
+    @Override
+    public String toString() {
+        return expression;
+    }
+
+    /**
+     * Parse the given expression
+     *
+     * @param expression - the expression to parse
+     * @throws ParseException - when the expression cannot be successfully be parsed
+     * @throws IllegalArgumentException - when expression parts conflict with each other
+     */
+    public void parseExpression(String expression) throws ParseException, IllegalArgumentException {
+
+        StringTokenizer expressionTokenizer = new StringTokenizer(expression, delimiters, false);
+        int position = 0;
+
+        setExpressionParts(new ArrayList<E>());
+        setCandidates(new ArrayList<Date>());
+
+        while (expressionTokenizer.hasMoreTokens()) {
+            String token = expressionTokenizer.nextToken().trim();
+            position++;
+            getExpressionParts().add(parseToken(token, position));
+        }
+
+        Collections.sort(getExpressionParts());
+
+        validateExpression();
+
+        if (startDate == null) {
+            setStartDate(Calendar.getInstance().getTime());
+        }
+
+        applyExpressionParts();
+
+        synchronized (this) {
+            continueSearch = true;
+            while (getCandidates().size() < minimumCandidates && continueSearch) {
+                populateWithSeeds();
+                getCandidates().clear();
+                applyExpressionParts();
+            }
+            continueSearch = false;
+        }
+
+        for (Date aDate : getCandidates()) {
+            logger.trace("Final candidate {} is {}", getCandidates().indexOf(aDate), aDate);
+        }
+    }
+
+    abstract protected void validateExpression() throws IllegalArgumentException;
+
+    protected void applyExpressionParts() {
+        Collections.sort(getExpressionParts());
+        for (ExpressionPart part : getExpressionParts()) {
+            logger.trace("Expanding {} from {} candidates", part.getClass().getSimpleName(), getCandidates().size());
+            setCandidates(part.apply(startDate, getCandidates()));
+            logger.trace("Expanded to {} candidates", getCandidates().size());
+            for (Date aDate : getCandidates()) {
+                logger.trace("Candidate {} is {}", getCandidates().indexOf(aDate), aDate);
+            }
+            prune();
```

Comment:
It is debatable, as some Expressions can generate a LOT of candidates. Just imagine one for every second in the Calendar. I just added it in so that the actual subclasses can decide how they manage the avalanche of candidate dates.